pkg/neotest/coverage.go

```diff
@@ -202,6 +209,42 @@ func documentSeqPoints(di *compiler.DebugInfo, doc documentName) []compiler.Debu
 	return res
 }
 
+// resolveOverlaps removes overlaps from debug points.
+// Its assumed that intervals can never overlap partially.
+func resolveOverlaps(points []compiler.DebugSeqPoint) []compiler.DebugSeqPoint {
+	var intervals []interval
+	for _, p := range points {
+		intervals = append(intervals, interval{DebugSeqPoint: p})
+	}
+	for i := range intervals {
+		for j := range intervals {
+			inner := &intervals[i]
+			outer := &intervals[j]
+			// If interval 'i' is already removed than there exists an even smaller interval that is also included by 'j'.
+			// This also ensures that if there are 2 equal intervals then at least 1 will remain.
+			if i == j || inner.remove {
+				continue
+			}
+			// outer interval start must be before inner interval start.
+			if !(outer.StartLine < inner.StartLine || outer.StartLine == inner.StartLine && outer.StartCol <= inner.StartCol) {
```

Comment:
> Also, I'm not quite sure why the first call to `runtime.CheckWitness` is marked as not covered, although it should be.

Pretty sure, its because there is no SeqPoint for it

Reply:
I have checked how Go itself handles statement coverage and it seems like it takes a line as a whole. So we could keep outer block if its one the same line as inner (both are one-liners), but remove outer block if its multi-line?

Reply:
Could you give an example of contract where coverage blocks are overlapping?

Reply:
`PutNamed` function in container contract.

Reply:
I'm finally back to this issue, let's locate the problem more precisely: consider `PutNamed` function of container contract and the coverage generated by master of NeoGo. I got this html coverage for the function:
![image](https://github.com/user-attachments/assets/05fe2903-c003-4e14-8daf-31e9e5912c91)
In the resulting coverage profile I got several batches of records for container contract:
[coverage.txt](https://github.com/user-attachments/files/17281335/coverage_init.txt)
The resulting blocks look like:
```

github.com/nspcc-dev/neofs-contract/contracts/container/contract.go:90.65,91.2 0 0
github.com/nspcc-dev/neofs-contract/contracts/container/contract.go:94.39,97.14 3 0
github.com/nspcc-dev/neofs-contract/contracts/container/contract.go:97.14,102.25 4 0
github.com/nspcc-dev/neofs-contract/contracts/container/contract.go:102.25,109.40 2 0
github.com/nspcc-dev/neofs-contract/contracts/container/contract.go:109.40,112.5 2 0
github.com/nspcc-dev/neofs-contract/contracts/container/contract.go:115.4,115.64 1 0
github.com/nspcc-dev/neofs-contract/contracts/container/contract.go:115.64,118.5 2 0
github.com/nspcc-dev/neofs-contract/contracts/container/contract.go:125.3,125.23 1 0
github.com/nspcc-dev/neofs-contract/contracts/container/contract.go:125.23,127.4 1 0
github.com/nspcc-dev/neofs-contract/contracts/container/contract.go:129.3,129.9 1 0
github.com/nspcc-dev/neofs-contract/contracts/container/contract.go:132.2,143.76 2 0
github.com/nspcc-dev/neofs-contract/contracts/container/contract.go:143.76,145.3 1 0
github.com/nspcc-dev/neofs-contract/contracts/container/contract.go:145.8,147.3 1 0
github.com/nspcc-dev/neofs-contract/contracts/container/contract.go:149.2,149.76 1 0
github.com/nspcc-dev/neofs-contract/contracts/container/contract.go:149.76,151.3 1 0
github.com/nspcc-dev/neofs-contract/contracts/container/contract.go:151.8,153.3 1 0
github.com/nspcc-dev/neofs-contract/contracts/container/contract.go:155.2,155.76 1 0
github.com/nspcc-dev/neofs-contract/contracts/container/contract.go:155.76,157.3 1 0
github.com/nspcc-dev/neofs-contract/contracts/container/contract.go:157.8,159.3 1 0
```
There are some duplicating records, in particular for `PutNamed`, but I suppose it's OK since multiple tests call `PutNamed` method.

Every record has the form of: `filename:StartLine.StartCol,EndLine.EndCol NumOfStmts ExecutionCnt`. Let's take the following record as an example:
```
github.com/nspcc-dev/neofs-contract/contracts/container/contract.go:254.21,259.87 4 0
```
This record covers the following selected contract statements:
```
![image](https://github.com/user-attachments/assets/147949e6-2616-485f-a3e5-45704acedd13)
```
First thing I should notice is that every record has `0` number of executions. It seems like a bug to me, but it's a separate issue that may be resolved in another PR.

Secondly, this record covers multiple statements (both in terms of code and in terms of VM instructions. Please, clarify is this a problem we're trying to solve? Do you think that coverage tool should produce a single record per a single code statement or per a single line of code?

Reply:
In vscode you can see that panic is not executed, but its "shadowed" by outside block. But in html it seems to be all green, which is worse.

Reply:
Attach VSCode result, please.